pr_agent/servers/azuredevops_server_webhook.py

```diff
@@ -0,0 +1,120 @@
+
+import json
+import os
+import re
+import secrets
+import uvicorn
+from fastapi import APIRouter, Depends, FastAPI, HTTPException
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+from fastapi.encoders import jsonable_encoder
+from starlette import status
+from starlette.background import BackgroundTasks
+from starlette.middleware import Middleware
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+from starlette_context.middleware import RawContextMiddleware
+
+from pr_agent.agent.pr_agent import PRAgent, command2class
+from pr_agent.config_loader import get_settings
+from pr_agent.log import get_logger
+from fastapi import Request, Depends
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+from pr_agent.log import get_logger
+
+security = HTTPBasic()
+router = APIRouter()
+available_commands_rgx = re.compile(r"^\/(" + "|".join(command2class.keys()) + r")\s*")
+azuredevops_server = get_settings().get("azure_devops_server")
+WEBHOOK_USERNAME = azuredevops_server.get("webhook_username")
+WEBHOOK_PASSWORD = azuredevops_server.get("webhook_password")
+
+def handle_request(
+    background_tasks: BackgroundTasks, url: str, body: str, log_context: dict
+):
+    log_context["action"] = body
+    log_context["api_url"] = url
+    with get_logger().contextualize(**log_context):
+        background_tasks.add_task(PRAgent().handle_request, url, body)
+
+
+# currently only basic auth is supported with azure webhooks
+# for this reason, https must be enabled to ensure the credentials are not sent in clear text
+def authorize(credentials: HTTPBasicCredentials = Depends(security)):
+        is_user_ok = secrets.compare_digest(credentials.username, WEBHOOK_USERNAME)
+        is_pass_ok = secrets.compare_digest(credentials.password, WEBHOOK_PASSWORD)
+        if not (is_user_ok and is_pass_ok):
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+                detail='Incorrect username or password.',
+                headers={'WWW-Authenticate': 'Basic'},
+            )
+        
+@router.post("/", dependencies=[Depends(authorize)])
+async def handle_webhook(background_tasks: BackgroundTasks, request: Request):
+    log_context = {"server_type": "azuredevops_server"}
+    data = await request.json()
+    get_logger().info(json.dumps(data))
+
+    actions = []
+    if data["eventType"] == "git.pullrequest.created": 
+        # API V1 (latest)
+        pr_url = data["resource"]["_links"]["web"]["href"].replace("_apis/git/repositories", "_git")
+        if get_settings().get("github_action_config").get("auto_review") == True:
+            actions.append("review")
+        if get_settings().get("github_action_config").get("auto_improve") == True:
+            actions.append("improve")
+        if get_settings().get("github_action_config").get("auto_describe") == True:
+            actions.append("describe")
```

Comment:
this is an inferior logic.

the logic of github_app.py is much better and richer, and enables any command to run

Reply:
here it allows any logic using comment command /, these run only the auto commands on PR creation

Reply:
should be fixed to github_app.py logic, which enable any command on startup